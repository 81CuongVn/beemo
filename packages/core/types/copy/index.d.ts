declare module 'copy' {
  // vinyl
  interface File {
    path: string;
  }

  interface Callback {
    (error: Error | null, files: File[]): void;
  }

  interface Options {
    cwd?: string;
    srcBase?: string;
  }

  export default function copy(
    patterns: string | string[] | object,
    dir: string,
    options?: Options | Callback,
    callback?: Callback,
  ): void;
}
