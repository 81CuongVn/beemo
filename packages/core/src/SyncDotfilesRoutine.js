/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import copy from 'copy';
import fs from 'fs-extra';
import path from 'path';
import { Routine } from 'boost';

export default class SyncDotfilesRoutine extends Routine {
  execute(): Promise<string[]> {
    this.task('Copying files', this.copyFilesFromConfigModule);
    this.task('Renaming files', this.renameFilesWithDot);

    return this.serializeTasks(this.context.configRoot);
  }

  /**
   * Copy all files from the config module's "dotfiles/" folder.
   */
  copyFilesFromConfigModule(configRoot: string): Promise<string[]> {
    const dotfilePath = path.join(configRoot, 'dotfiles/*');

    return new Promise((resolve, reject) => {
      copy(dotfilePath, this.tool.options.root, (error, files) => {
        this.tool.invariant(
          !error,
          `Coping dotfiles from ${chalk.cyan(dotfilePath)}`,
          'Copied',
          'Failed',
        );

        if (error) {
          reject(error);
        } else {
          resolve(files.map((file) => {
            this.tool.emit('create-dotfile', [file.path]);

            this.tool.debug(`  ${chalk.gray(file.path)}`);

            return file.path;
          }));
        }
      });
    });
  }

  /**
   * The original files are not prefixed with ".", as it causes git/npm issues
   * in the repository. So we need to rename them after they are copied.
   */
  renameFilesWithDot(filePaths: string[]): Promise<string[]> {
    this.tool.debug('Renaming dotfiles and prefixing with a period');

    return Promise.all(filePaths.map((filePath) => {
      const dir = path.dirname(filePath);
      const newName = `.${path.basename(filePath)}`;
      const newPath = path.join(dir, newName);

      return fs.rename(filePath, newPath).then(() => {
        this.tool.emit('rename-dotfile', [newPath]);

        this.tool.log(`${chalk.gray('->')} ${newName}`);

        this.tool.debug(`  ${chalk.gray(newPath)}`);

        return newPath;
      });
    }));
  }
}
