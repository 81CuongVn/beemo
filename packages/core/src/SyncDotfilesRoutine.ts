/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import copy from 'copy';
import fs from 'fs-extra';
import path from 'path';
import { Routine } from 'boost';
import optimal, { string, Struct } from 'optimal';
import isPatternMatch from './utils/isPatternMatch';
import { Context } from './types';

export interface SyncDotfilesOptions extends Struct {
  filter: string;
}

export default class SyncDotfilesRoutine extends Routine<SyncDotfilesOptions, Context> {
  bootstrap() {
    this.options = optimal(
      this.options,
      {
        filter: string().empty(),
      },
      {
        name: 'SyncDotfilesRoutine',
      },
    );
  }

  execute(context: Context): Promise<string[]> {
    this.task('Copying files', this.copyFilesFromConfigModule);
    this.task('Renaming files', this.renameFilesWithDot);

    return this.serializeTasks(context.moduleRoot);
  }

  /**
   * Copy all files from the config module's "dotfiles/" folder.
   */
  copyFilesFromConfigModule(context: Context, moduleRoot: string): Promise<string[]> {
    const dotfilePath = path.join(moduleRoot, 'dotfiles/*');
    const { filter } = this.options;

    return new Promise((resolve, reject) => {
      copy(dotfilePath, this.tool.options.root, (error, files) => {
        this.debug.invariant(
          !error,
          `Coping dotfiles from ${chalk.cyan(dotfilePath)}`,
          'Copied',
          'Failed',
        );

        if (filter) {
          this.debug('Filtering dotfiles with "%s"', filter);

          files = files.filter(file => isPatternMatch(file.path, filter));
        }

        if (error) {
          reject(error);
        } else {
          resolve(
            files.map(file => {
              this.tool.emit('copy-dotfile', [file.path]);

              this.debug('  %s', chalk.gray(file.path));

              return file.path;
            }),
          );
        }
      });
    });
  }

  /**
   * The original files are not prefixed with ".", as it causes git/npm issues
   * in the repository. So we need to rename them after they are copied.
   */
  renameFilesWithDot(context: Context, filePaths: string[]): Promise<string[]> {
    this.debug('Renaming dotfiles and prefixing with a period');

    return Promise.all(
      filePaths.map(filePath => {
        const dir = path.dirname(filePath);
        const newName = `.${path.basename(filePath)}`;
        const newPath = path.join(dir, newName);

        return fs.rename(filePath, newPath).then(() => {
          this.tool.emit('rename-dotfile', [newPath]);

          this.tool.log('%s %s', chalk.gray('->'), newName);

          this.debug('  %s', chalk.gray(newPath));

          return newPath;
        });
      }),
    );
  }
}
