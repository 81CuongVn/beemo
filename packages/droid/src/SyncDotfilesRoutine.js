/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import copy from 'copy';
import fs from 'fs-extra';
import path from 'path';
import { Routine } from 'boost';

export default class SyncDotfilesRoutine extends Routine {
  execute(configRoot: string): Promise<string[]> {
    this.task('Copying files', this.copyFilesFromConfigModule);
    this.task('Renaming files', this.renameFilesWithDot);

    return this.serializeTasks(configRoot);
  }

  /**
   * Copy all files from the config module's "dotfiles/" folder.
   */
  copyFilesFromConfigModule(configRoot: string): Promise<string[]> {
    const dotfilePath = path.join(configRoot, 'dotfiles/*');

    return new Promise((resolve, reject) => {
      copy(dotfilePath, this.tool.options.root, (error, files) => {
        this.tool.invariant(
          !error,
          `Coping dotfiles from ${chalk.cyan(dotfilePath)}`,
          'Copied',
          'Failed',
        );

        if (error) {
          reject(error);
        } else {
          resolve(files.map((file) => {
            this.tool.emit('create-dotfile', null, [file.path]);

            this.tool.debug(`\t${file.path}`);

            return file.path;
          }));
        }
      });
    });
  }

  /**
   * The original files are not prefixed with ".", as it causes git/npm issues
   * in the repository. So we need to rename them after they are copied.
   */
  renameFilesWithDot(filePaths: string[]): Promise<string[]> {
    this.tool.debug('Renaming dotfiles and prefixing with a period');

    return Promise.all(filePaths.map((filePath) => {
      const dir = path.dirname(filePath);
      const newName = `.${path.basename(filePath)}`;
      const newPath = path.join(dir, newName);

      // TODO show from config module as prefix?
      this.tool.log(`-> ${newName}`);

      return fs.rename(filePath, newPath).then(() => newPath);
    }));
  }
}
