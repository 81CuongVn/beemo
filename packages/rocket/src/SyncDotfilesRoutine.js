/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import copy from 'copy';
import fs from 'fs-extra';
import path from 'path';
import { Routine } from 'boost';

export default class SyncDotfilesRoutine extends Routine {
  execute(configRoot: string): Promise<string[]> {
    this.task('Copying files', this.copyFilesFromConfigModule);
    this.task('Renaming files', this.renameFilesWithDot);

    return this.serializeTasks(configRoot);
  }

  /**
   * Copy all files from the config module's "dotfiles/" folder.
   */
  copyFilesFromConfigModule(configRoot: string): Promise<string[]> {
    return new Promise((resolve, reject) => {
      copy(path.join(configRoot, 'dotfiles/*'), this.tool.options.root, (error, files) => {
        if (error) {
          reject(error);
        } else {
          resolve(files.map(file => file.path));
        }
      });
    });
  }

  /**
   * The original files are not prefixed with ".", as it causes git/npm issues
   * in the repository. So we need to rename them after they are copied.
   */
  renameFilesWithDot(filePaths: string[]): Promise<string[]> {
    return Promise.all(filePaths.map((filePath) => {
      const dir = path.dirname(filePath);
      const newName = `.${path.basename(filePath)}`;
      const newPath = path.join(dir, newName);

      // TODO show from config module as prefix?
      this.tool.log(`-> ${newName}`);

      return fs.rename(filePath, newPath).then(() => newPath);
    }));
  }
}
