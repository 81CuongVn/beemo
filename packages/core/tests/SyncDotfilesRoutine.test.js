import { Tool } from 'boost';
import copy from 'copy';
import fs from 'fs-extra';
import SyncDotfilesRoutine from '../src/SyncDotfilesRoutine';
import { createContext, setupMockTool, prependRoot, getRoot } from '../../../tests/helpers';

jest.mock('copy');
jest.mock('fs-extra');
jest.mock('boost/lib/Tool');

describe('SyncDotfilesRoutine', () => {
  let routine;

  beforeEach(() => {
    routine = new SyncDotfilesRoutine('sync', 'Syncing dotfiles');
    routine.context = createContext();
    routine.tool = setupMockTool(new Tool());
    routine.debug = jest.fn();
    routine.debug.invariant = jest.fn();

    copy.mockImplementation((filePath, root, callback) => {
      callback(null, [{ path: './foo' }, { path: './bar' }, { path: './baz' }]);
    });

    fs.rename.mockImplementation(value => Promise.resolve(value));
  });

  describe('bootstrap()', () => {
    it('errors if filter is not a string', () => {
      routine.options.filter = 123;

      expect(() => {
        routine.bootstrap();
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('execute()', () => {
    it('passes module root to tasks', () => {
      routine.serializeTasks = jest.fn();
      routine.execute(routine.context);

      expect(routine.serializeTasks).toHaveBeenCalledWith(getRoot());
    });

    it('executes pipeline in order', async () => {
      const copySpy = jest.spyOn(routine, 'copyFilesFromConfigModule');
      const renameSpy = jest.spyOn(routine, 'renameFilesWithDot');

      const paths = await routine.execute(routine.context);

      expect(copySpy).toHaveBeenCalledWith(routine.context, getRoot(), expect.anything());
      expect(renameSpy).toHaveBeenCalledWith(
        routine.context,
        ['./foo', './bar', './baz'],
        expect.anything(),
      );
      expect(paths).toEqual(['.foo', '.bar', '.baz']);
    });
  });

  describe('copyFilesFromConfigModule()', () => {
    it('calls it with correct path arguments', async () => {
      await routine.copyFilesFromConfigModule(routine.context, getRoot());

      expect(copy).toHaveBeenCalledWith(prependRoot('dotfiles/*'), getRoot(), expect.anything());
    });

    it('returns file paths as strings', async () => {
      const paths = await routine.copyFilesFromConfigModule(routine.context, './root');

      expect(paths).toEqual(['./foo', './bar', './baz']);
    });

    it('filters files using config', async () => {
      routine.options.filter = 'a(r|z)';

      const paths = await routine.copyFilesFromConfigModule(routine.context, './root');

      expect(paths).toEqual(['./bar', './baz']);
    });

    it('handles errors', async () => {
      copy.mockImplementation((filePath, root, callback) => callback(new Error('Oops')));

      try {
        await routine.copyFilesFromConfigModule(routine.context, './root');
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
      }
    });

    it('triggers `copy-dotfile` event', async () => {
      const spy = routine.tool.emit;

      await routine.copyFilesFromConfigModule(routine.context, './root');

      expect(spy).toHaveBeenCalledWith('copy-dotfile', ['./foo']);
      expect(spy).toHaveBeenCalledWith('copy-dotfile', ['./bar']);
      expect(spy).toHaveBeenCalledWith('copy-dotfile', ['./baz']);
    });
  });

  describe('renameFilesWithDot()', () => {
    it('renames each file path', async () => {
      const paths = await routine.renameFilesWithDot(routine.context, [
        'foo',
        './path/bar',
        '/path/baz',
      ]);

      expect(paths).toEqual(['.foo', 'path/.bar', '/path/.baz']);

      expect(fs.rename).toHaveBeenCalledWith('foo', '.foo');
      expect(fs.rename).toHaveBeenCalledWith('./path/bar', 'path/.bar');
      expect(fs.rename).toHaveBeenCalledWith('/path/baz', '/path/.baz');
    });

    it('triggers `rename-dotfile` event', async () => {
      const spy = routine.tool.emit;

      await routine.renameFilesWithDot(routine.context, ['foo', './path/bar', '/path/baz']);

      expect(spy).toHaveBeenCalledWith('rename-dotfile', ['.foo']);
      expect(spy).toHaveBeenCalledWith('rename-dotfile', ['path/.bar']);
      expect(spy).toHaveBeenCalledWith('rename-dotfile', ['/path/.baz']);
    });
  });
});
