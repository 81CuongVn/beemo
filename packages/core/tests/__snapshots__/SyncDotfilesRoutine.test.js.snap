// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncDotfilesRoutine bootstrap() errors if filter is not a string 1`] = `"Invalid SyncDotfilesRoutine field \\"filter\\". Must be a string."`;
